### PR TITLE
mempool: return mempool errors to the abci client

### PR DESCRIPTION
**Pull request by williambanfield**
This changes adds an `MempoolError` field to the `ResponseCheckTx`. This will allow clients to understand that their transaction was rejected from the mempool despite passing the ABCI check. 

This change also updates the code to make use of early returns to prevent highly nested code blocks. Namely, it returns when the type assertion fails at the beginning of the method, instead of wrapping the entire method in a large if statement. This has a somewhat large effect on the diff as rendered by github.

addresses: #3546

### CODE REVIEW

File: internal/mempool/v1/mempool.go [hunk 1]
```diff
@@ -504,104 +504,109 @@ func (txmp *TxMempool) Update(
...
+		// ignore bad transactions
+		txmp.logger.Info(
+			"rejected bad transaction",
+			"priority", wtx.priority,
+			"tx", fmt.Sprintf("%X", wtx.tx.Hash()),
+			"peer_id", txInfo.SenderNodeID,
+			"code", checkTxRes.CheckTx.Code,
+			"post_check_err", err,
+		)
+		checkTxRes.CheckTx.MempoolError.ErrorMessage = err.Error()
```
Comment (alexanderbez) [inline internal/mempool/v1/mempool.go:526]: err could be nil here

File: proto/tendermint/abci/types.proto [hunk 2]
```diff
@@ -185,17 +185,23 @@ message ResponseBeginBlock {
...
+  bytes          data          = 2;
+  string         log           = 3;  // nondeterministic
+  string         info          = 4;  // nondeterministic
+  int64          gas_wanted    = 5 [json_name = "gas_wanted"];
+  int64          gas_used      = 6 [json_name = "gas_used"];
+  repeated Event events        = 7 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "events,omitempty"];
+  string         codespace     = 8;
+  string         sender        = 9;
+  int64          priority      = 10;
+  MempoolError   mempool_error = 11;
```
Comment (alexanderbez) [inline proto/tendermint/abci/types.proto:204]:
```suggestion
  MempoolError   mempool_error = 11;
```

What are your thoughts on having this just be an `error`, i.e. a `string`?

Comment (williambanfield) [reply proto/tendermint/abci/types.proto:204]: I think it depends on how we expect the client to handle the error. If we expect the client to just log the error and do nothing, then I think it makes sense to have the error be just a string. If the client should react to the error in a programmatic way, I think we should have metadata that helps the client decide what to do about the error. 

Do you have any sense for how the client will use this error?

Comment (alexanderbez) [reply proto/tendermint/abci/types.proto:204]: > I think it depends on how we expect the client to handle the error. If we expect the client to just log the error and do nothing, then I think it makes sense to have the error be just a string.

This is exactly how it'll happen. The response will come back to the client and it'll be printed in the CLI or returned as an API response. The client will see "something about gas" and will either try again or move on.

The client will not react in a programmatic way to this AFAIK.

Comment (williambanfield) [reply proto/tendermint/abci/types.proto:204]: Great. Then no need to fuss over that. I'll move that field to a string, put up tests for this change and move it out of draft.

**Comment by williambanfield** [comment]
I think this is pretty much there, but I'm not sure quite how to test the RPC changes. It seems challenging to provide a mocked mempool to the node created in the RPC tests.

**Comment by alexanderbez** [comment]
I wouldn't worry about testing the RPC-part of this changeset. I would just test and ensure the mempool is populating the field in ResponseCheckTx, which is a mempool test and should be pretty straightforward to do I think?

File: proto/tendermint/abci/types.proto [hunk 3]
```diff
@@ -186,16 +186,17 @@ message ResponseBeginBlock {
...
+  bytes          data                  = 2;
+  string         log                   = 3;  // nondeterministic
+  string         info                  = 4;  // nondeterministic
+  int64          gas_wanted            = 5 [json_name = "gas_wanted"];
+  int64          gas_used              = 6 [json_name = "gas_used"];
+  repeated Event events                = 7 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "events,omitempty"];
+  string         codespace             = 8;
+  string         sender                = 9;
+  int64          priority              = 10;
+  string         mempool_error_message = 11;
```
Comment (alexanderbez) [inline proto/tendermint/abci/types.proto:199]: nit:

```suggestion
  string         mempool_error = 11;
```

File: rpc/core/types/responses.go [hunk 4]
```diff
@@ -187,10 +187,11 @@ type ResultConsensusState struct {
...
 type ResultBroadcastTx struct {
-	Code      uint32         `json:"code"`
-	Data      bytes.HexBytes `json:"data"`
-	Log       string         `json:"log"`
-	Codespace string         `json:"codespace"`
+	Code                uint32         `json:"code"`
+	Data                bytes.HexBytes `json:"data"`
+	Log                 string         `json:"log"`
+	Codespace           string         `json:"codespace"`
+	MempoolErrorMessage string         `json:"mempool_error_message"`
```
Comment (alexanderbez) [inline rpc/core/types/responses.go:194]: nit:

```suggestion
	MempoolError string         `json:"mempool_error"`
```

File: rpc/core/mempool.go [hunk 5]
```diff
@@ -48,11 +48,12 @@ func (env *Environment) BroadcastTxSync(ctx *rpctypes.Context, tx types.Tx) (*ct
...
-		Log:       r.Log,
-		Codespace: r.Codespace,
-		Hash:      tx.Hash(),
+		Code:                r.Code,
+		Data:                r.Data,
+		Log:                 r.Log,
+		Codespace:           r.Codespace,
+		MempoolErrorMessage: r.MempoolErrorMessage,
+		Hash:                tx.Hash(),
 	}, nil
```
Comment (alexanderbez) [inline rpc/core/mempool.go:57]: We also need to set `MempoolError` during `BroadcastTxCommit` since that calls CheckTx as well (around line 107).

**Review by alexanderbez — CHANGES_REQUESTED** [review]
Looks good! We're just missing the commit case.

File: rpc/core/mempool.go (hunk 5 above)
Comment (williambanfield) [reply rpc/core/mempool.go:57]: I think it uses the result of `initTxCallback` directly. The code in BroadcastTxCommit calls `CheckTx` [with a callback to capture the response](https://github.com/tendermint/tendermint/blob/aacfc596993f46c0d90e43753116908094b5cb15/rpc/core/mempool.go#L40 ). This code is executed [after](https://github.com/tendermint/tendermint/blob/aacfc596993f46c0d90e43753116908094b5cb15/internal/mempool/v1/mempool.go#L287) the response is received and receives the entire response modified from `initTxCallback`. As a result, I think the field should already be set on the response we send.

**Review by alexanderbez — APPROVED** [review]

**Comment by williambanfield** [comment]
Thanks for the review @alexanderbez ! 🙏🙏🙏

**Comment by williambanfield** [comment]
> Looks good! We're just missing the commit case.

Thanks for the review!

File: proto/tendermint/abci/types.proto [hunk 6]
```diff
@@ -186,16 +186,17 @@ message ResponseBeginBlock {
...
+  bytes          data          = 2;
+  string         log           = 3;  // nondeterministic
+  string         info          = 4;  // nondeterministic
+  int64          gas_wanted    = 5 [json_name = "gas_wanted"];
+  int64          gas_used      = 6 [json_name = "gas_used"];
+  repeated Event events        = 7 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "events,omitempty"];
+  string         codespace     = 8;
+  string         sender        = 9;
+  int64          priority      = 10;
+  string         mempool_error = 11;
```
Comment (marbar3778) [inline proto/tendermint/abci/types.proto:199]: This change needs to be reflected in the spec. 

Also can you add a comment, does the application populate this field or should it be ignored?

This PR needs a changelog!

**Review by marbar3778 — COMMENTED** [review]
changelog entry please

**Comment by alexanderbez** [comment]
@williambanfield can you please create another PR for a quick changelog entry. Sorry, I missed this in review @marbar3778.

**Comment by williambanfield** [comment]
@alexanderbez On it. Thanks for the catch @marbar3778.

**Pull request merged**